sensor/kubernetes/certrefresh/tls_issuer_common.go

```diff
@@ -35,3 +50,187 @@ type certificateRefresherGetter func(certsDescription string, requestCertificate
+}
+
+// ProcessMessage dispatches Central's messages to Sensor received via the Central receiver.
+// This method must not block as it would prevent centralReceiverImpl from sending messages
+// to other SensorComponents.
+func (i *tlsIssuerImpl) ProcessMessage(msg *central.MsgToSensor) error {
+	if i.getResponseFn == nil {
+		return errors.New("getResponseFn is not set")
+	}
+
+	response := i.getResponseFn(msg)
+	if response == nil {
+		// messages not supported by this component are ignored
+		return nil
+	}
+
+	go i.dispatch(response)
+	return nil
+}
+
+func (i *tlsIssuerImpl) dispatch(response *Response) {
+	if !i.requestOngoing.Load() {
+		log.Warnf("Received a response for request ID %q, but there is no ongoing RequestCertificates call.", response.RequestId)
+		return
+	}
+	ongoingRequestID := i.getOngoingRequestID()
+	if ongoingRequestID != response.RequestId {
+		log.Warnf("Request ID %q does not match ongoing request ID %q.",
+			response.RequestId, ongoingRequestID)
+		return
+	}
+
+	i.responseFromCentral.Store(response)
+	i.responseReceived.Signal()
+}
+
+// requestCertificates makes a new request for a new set of secured cluster certificates from Central.
+// Concurrent requests are *not* supported.
+func (i *tlsIssuerImpl) requestCertificates(ctx context.Context) (*Response, error) {
+	if i.requiredCentralCapability != nil {
+		// Central capabilities are only available after this component is created,
+		// which is why this check is done here
+		if !centralcaps.Has(*i.requiredCentralCapability) {
+			return nil, fmt.Errorf("TLS certificate refresh failed: missing Central capability %q", *i.requiredCentralCapability)
+		}
+	}
+
+	if !i.requestOngoing.CompareAndSwap(false, true) {
+		return nil, errors.New("concurrent requests are not supported.")
+	}
+
+	defer func() {
+		i.requestOngoing.Store(false)
+		i.responseReceived.Reset()
+	}()
+
+	requestID := uuid.NewV4().String()
+	i.setOngoingRequestID(requestID)
+
+	if err := i.send(ctx, requestID); err != nil {
+		return nil, err
+	}
+	return i.receive(ctx)
+}
+
+// send a cert refresh request to Central
+func (i *tlsIssuerImpl) send(ctx context.Context, requestID string) error {
+	select {
+	case <-i.stopSig.Done():
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	case i.msgToCentralC <- message.New(i.newMsgFromSensorFn(requestID)):
+		return nil
+	}
+}
+
+// receive handles the response to a specific certificate request
+func (i *tlsIssuerImpl) receive(ctx context.Context) (*Response, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-i.responseReceived.Done():
+		return i.responseFromCentral.Load(), nil
+	}
+}
+
+func (i *tlsIssuerImpl) setOngoingRequestID(requestID string) {
+	i.ongoingRequestIDMutex.Lock()
+	defer i.ongoingRequestIDMutex.Unlock()
+
+	i.ongoingRequestID = requestID
+}
+
+func (i *tlsIssuerImpl) getOngoingRequestID() string {
+	i.ongoingRequestIDMutex.Lock()
+	defer i.ongoingRequestIDMutex.Unlock()
+
+	return i.ongoingRequestID
+}
```

Comment:
Those two funcs are totally cool, but we also have this construct that may be useful (more esthetic) in some cases: https://github.com/stackrox/stackrox/blob/4430f5c979ef4dbf84f9ba97e8192d2d4bf57b6e/pkg/concurrency/lock.go#L29

Reply:
Thanks, done!